lib/src/clojure_lsp/feature/code_actions.clj

```diff
@@ -300,6 +308,9 @@
       @can-cycle-fn-literal?*
       (conj (cycle-fn-literal-action uri line character))
 
+      @can-promote-fn?*
+      (conj (promote-fn-action uri line character))
+
```

Comment:
I wonder if those 2 could cause questions what's the difference between then when on a `#()`, since both do the same, right?
Why `promote function` couldn't move `#()` -> `defn` directly? and for the other case we could use cycle fn literal, WDYT?

Reply:
Yeah, that's true, they both do the same thing. I suppose that is confusing. I think you had proposed `anything` -> `defn` earlier. OK, I'll experiment with that. It's a little harder but I'll see what I can figure out.

Reply:
Another way to resolve the confusion would be to leave `promote-fn` as is, and replace `cycle-fn-literal` with `demote-fn`. So:

* `promote-fn`: `#()` -> `fn` -> `defn`
* `demote-fn`: `fn` -> `#()`

I think either is possible, so it's just what you think is clearest.

Reply:
Yeah, your suggestion sounds good enough, as we would have clearer code action names for each operation.
Since cycle-fn-literal is quite new, I'm ok of just renaming it too, as a breaking change

Reply:
+1 . At some future point it might be nice for `demote-fn` to inline defn too: `defn` -> `fn` -> `#()`, then `promote-fn` would look for all matching `fn` in the project and replace with call to def'd var.

```clojure
[(map |#(inc %) xs) (map #(inc %) ys)]
;; promote =>
(defn myinc [x] (inc x))
[(map |myinc xs) (map myinc ys)]
;; demote =>
[(map |#(inc %) xs) (map #(inc %) ys)]
```

Reply:
> Since cycle-fn-literal is quite new, I'm ok of just renaming it too, as a breaking change

👍

Reply:
> At some future point it might be nice for demote-fn to inline defn too: defn -> fn

I was thinking that'd be a nice enhancement too. Someday.

Reply:
@snoe your examples are a little confusing, and I want to make sure I'm not missing anything. The `demote` part makes sense. But the `promote` examples aren't really in line with how `promote-fn` is working in this PR. I'm not sure if that's because you're showing pseudocode or because you expect `promote-fn` to behave differently.

The first difference is that I don't plan to make `promote-fn` convert `#()` directly to `defn`, though we talked about that at one point. Instead, it'll work more like how clj-refactor worked, `#() -> fn -> defn`. 

Second, it won't replace every identical occurrence, which is what your example looks like it's suggesting. Perhaps it could, but IIRC, that's not how clj-refactor worked. Also I think it'd be very surprising, since it could replace occurrences several screens away.

Are we on the same page?

Reply:
yeah i skipped the fn step for brevity.

This is all in the someday enhancement. if demote inlines defn, promote should do the inverse, which I believe would have the effect you describe. Maybe both only operate on private defn to minimize the surface area. Maybe it would be too surprising and be a bad idea, but in concept i think i like symmetry.

Reply:
I see. Thanks for the clarification. The symmetry is appealing. And I loved clj-refactor's `expand-let` feature where it [replaced identical entries](https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-expand-let). (Wouldn't it be nice if clojure-lsp did that?) But `let`s are much more local than `defn`s. Anyway, something to ponder in the future.